docs/extending-nsm.md

```diff
@@ -0,0 +1,71 @@
+# Extending Network Service Mesh (NSM)
+
+
+# Introduction
+
+This document is initially aimed at developers looking to extend the NSM with new plugins to provide users with new functionality. It is based on the version of NSM available August 2018 in the [Network Service Mesh Github](https://github.com/ligato/networkservicemesh) Repository. It documents the steps that were used to create the simple-dataplane extension. 
+
+All steps assume the top level directory is networkservicemesh from github.
+
+
+# Development Steps
+
+The development steps are to add a new data plane called cnf-dataplane.
+
+
+## Edit Makefile
+
+The top level Makefile [networkservicemesh/Makefile] to add to the following targets
+
+docker-build: add "docker-build-cnf-dataplane"
+
+docker-push: add "docker-push-cnf-dataplane"
+
+
+## Dockerfile
+
+A new dockerfile is added: networkservicemesh/build/Dockerfile-cnf.dataplane
+
+
+## API Directory and Interface Definition
+
+Create a new directory for the cnf-dataplane daemonset CRD interface definition.
+
+The new directory is: networkservicemesh/pkg/nsm/apis/cnfdataplane
+
+The interface definition file written in Google Protobufs is added to this directory. 
+
+The file will be:  networkservicemesh/pkg/nsm/apis/cnfdataplane/cnfdataplane.proto
```

Comment:
It would be great instead of telling plugin developer to create new API and us ending up with 10k of different APIs, suggest plugin developer to check already existing api.  Ideally we should propose the common API for dataplane which should cover 90% of requirements for dataplane plugins, otherwise we might end in unmanageable situation. 
The reason the common API has not been yet proposed is because of the idea to use VPP as a base in terms of requirements.

Reply:
One of the things I would like to figure out how to do is put all the changes in a separate directory structure. This does not exactly solve the issue but it keeps all changes for a plugin in a single directory and users do not see the APIs they do not care about.  Also we could then only build individual plugins etc.